### ISSUE TITLE
All games have graphical corruption with Vulkan on RetroArch

### ISSUE THREAD

**Issue by PoloniumRain** [flyinghead/flycast#1631]
Platform / OS / Hardware: 

RetroArch 1.19.1 on PC, Windows 11, AMD Threadripper 3960X, RTX 3080 (latest drivers)

**_Description of the Issue_**

Graphical corruption in the form of many small flickering blocks, shown in the videos below. I've tried 12 random games so far and it affects each one. RetroArch Shaders have an effect on it, but it also always happens **without** shaders. So Vulkan is unusable because this problem is always visable. Depending on the game and area it will sometimes have less or more graphical corruption when a shader is either enabled or disabled. I also have the D3D11 shader problem mentioned in #1612, so GL is only thing that's working for me.

**_Debugging Steps Tested_**

  * Clean install of RetroArch with default settings.
  * Default Flycast core settings.
  * I've also tried many different core settings. None have any effect.
  * D3D11 has graphical corruption but only with shaders enabled.
  * GL works fine with or without shaders.
  * Older Nvidia drivers from 5+ months ago also have this issue
  

https://github.com/user-attachments/assets/45656b2e-4b48-4629-bcaa-121cbebe7981

https://github.com/user-attachments/assets/a0ec1dd5-2ff0-4677-9917-8ab651de255f

https://github.com/user-attachments/assets/7271c5fd-5990-4f3e-ba76-c1be03bfbc6f

**Comment by flyinghead** [comment]
This is very likely a GPU or GPU driver issue.
Does it happen only in Per Pixel Transparent Sorting? Or also in Per Triangle and Per Strip?

**Comment by flyinghead** [comment]
Can you try Flycast standalone?

**Comment by PoloniumRain** [comment]
Just tried 8 games with standalone v2.3 and Vulkan works fine. 

I've also now tried 20+ games with the RetroArch Flycast core and they all have the same graphical corruption, including the Dreamcast menu that appears when there's no disc in the console.

**Comment by flyinghead** [comment]
I would try to do a fresh install of retroarch, using default settings and no add-ons, shaders, ... and see if the issue persists.
Trying older RA versions could also help determine if it's a regression in 1.19.1.

**Comment by PoloniumRain** [comment]
> I would try to do a fresh install of retroarch, using default settings and no add-ons, shaders

Already tried that. 

But i've now tested older RetroArch versions, going back to 1.14  Theses all work without any issues with Vulkan. The problem was introduced in version 1.19.0, which was released on the 30th of May. So i'll make a post about this issue on the RA GitHub...

**Comment by CO-Windler** [comment]
On my Ryzen 2400G APU Vulkan also has glitches (e.g. water pixel mess in "Soul Surfer"). DirectX11 works better (only Sega Driving Simulator depends on OpenGL, which seems slower/choppier on my APU).

**Comment by cdf83** [comment]
Not sure if related but I find recently that Sega Rally 2 freezes quite often in RetroArch, particularly in the menus after the name entry screen. It doesn't seem to do this in standalone. 

It was more frequent with the gl driver but also occurs with d3d11

**Comment by ToniBC** [comment]
The corruption appears even in the Flycast 2.3.2-199 nightly and in Retroarch with the Vulkan Driver. Even in the BIOS as they say.

With DX11 it seems to be mitigated or it doesn't appear, the same if you set the native resolution, it doesn't appear in the BIOS, but if you increase the resolution, there are artifacts in the BIOS at the bottom right, above the clock icon, both in Flycast and Retroarch.

If Full Framebuffer Emulation is activated, the artifacts no longer appear, but we can't increase the resolution.

This has been happening for a while and as it happens in both, it's not focused on AR.

Many won't notice it because they use old versions, like March 2.3, but it seems to be something related to the framebuffer.

**Comment by flyinghead** [comment]
@ToniBC if you increase the resolution you might need to increase the pixel buffer size. What you're seeing may be the result othe pixel buffer overflowing.
So far this issue only happens on the flycast core, not standalone.

**Comment by ToniBC** [comment]
This one has 128 layers and 4GB of pixel buffer. Tested by default and the same thing happens.

I have tested the same configuration (even moving the .cfg from one version to the other) with Flycast 2.3 and it doesn't cause any problems, those errors are not observed in the bios or in the games, but the same thing happens with the latest versions.

I will try to test other previous versions to see if the same thing happens.

**Comment by ToniBC** [comment]
I have done several tests with several versions, I'll tell you.

In version 2.3.2-4 (2024-05-21) everything apparently works fine, the glitch does not appear in the bios.

In version 2.3.2-95 (2024-05-23), the glitch already appears there. The changes introduced there, some broke something. I'll put the changes here.

* Merge remote-tracking branch 'origin/dev'
* android: update project
* deps: update sdl to version 2.30.3
* deps: update libchdr
* deps: update imgui to version 1.90.6
* cmake: better version conversion for uwp
* Always use BSD sed
* dx11: use ALLOW_TEARING flag to disable vsync if available
* gl: only display crosshairs when needed
* gl: refactor vmu and xhair drawing. Blit rendered frame only once
* std::swap is undefined for vector
* vk: multi render support
* fix compile warning
* naomi: fix marstv freeze
* os_notify
* ui: load savestate thumbnails asynchronously. limit thumbnail size
* ios build fix
* switch, uwp, libretro build fixes
* save screenshot. add screenshot to savestates
* achievements: fix size of multiple challenge indicators popup
* ui: bug fixes. cheats window changes
* work around mingw-w64 init crash
* move ui to its own folder
* achievements: leaderboard notifications. detailed toast message
* ui: split display_settings
* ui: use uiScaled()
* ui: use imgui for vmu on pause menu. toast notifications
* ui: crash when loading asian fonts on windows
* achievements: better game load message when no achievements
* ui: change pause menu layout. add icons here and there
* vk: don't delete the ImGui driver when recreating the swap chain
* achievements: hardcore mode
* achievements: fix crash when loading image cache
* ui: achievements list, new pause menu
* Use Xcode 15.3
* better RetroAchievements UI and threading
* move http_client to oslib
* macos: bump minimum macOS version to 10.11 to avoid build error
* Merge remote-tracking branch 'origin/master' into dev
* RetroAchievements support
* ci: update netbsd and openbsd versions
* savestate: drop support for legacy libretro and * move most os_* funcs to oslib
* lr: switch build fix
* drop dispmanx support. move switch main() to its own file
* get rid of os_GetSeconds()
* lr win32 build fix
* vk: move macOS video routing code from context to renderer
* minor input and lua fixes
* set thread names
* macOS build fix
* thegamedb: ignore common disk ids T0000 and T0000M
* ui: rearrange Settings > Video screen
* android workflow: fetch submodules recursively (for libadrenotools)
* vulkan: allow custom GPU driver loading with libadrenotools
* maple: implement AllStatusReq for controller and VMU
* msvc build fix
* Fix vmu beep frequency
* bump libchdr to match master
* [Libretro] Check for per-pixel compatibility and hide the option if not supported
* Merge remote-tracking branch 'origin/master' into dev
* richer discord presence using boxart from thegamesdb
* fix compile warnings
* Discord Presence support
* set currently played game name in the window title
* android: hid barcode scanner support
* fix FreeBSD x64 segfault context
* isofs: fix long directory listing
* bump depth scale for NHL 2K2
* sh4: fix P4 region mapping and on chip ram addressing
* deps: sync imgui_impl_vulkan.cpp with upstream
* deps: update imgui to version 1.90.4
* sdl: support multiple mice when allowed by the platform
* naomi: fix shaktam inputs
* cmake: improved parallelism in msbuild
* ci: use cross-platform-actions/action to build on bsd systems
* bsd: update segfault context for arm64
* deps: update imgui to version 1.90.3
* linux: use egl with x11 instead of glx
* deps: update libchdr
* deps: update libzip to version 1.10.1
* ci: use latest devkitpro docker image
* init drive when running .elf. Skip special settings for BIOS and .elf.
* gladLoader: check return code and log. cheats: light refactoring
* Merge remote-tracking branch 'origin/master' into dev
* samsptk: use cheat instead of extra depth scale
* gdrom: delay "no disk" reporting to let the BIOS play the boot animation
* pvr: apply a negative bias to background plane depth
* ui: replace Exit by Close Game. Better format for some UI values.
* gdrom: set correct lead-out FAD of single-density area

**Comment by flyinghead** [comment]
This is a merge commit from the dev branch, which includes many changes.
Can you try to same exercise on the dev branch to narrow it down a bit?

**Comment by ToniBC** [comment]
Unfortunately I don't have enough knowledge to compile versions, the ones I found were those, between those two I don't have more versions. The oldest buildbot is from 2024-07 which is where the bug is already in the master branch.

**Comment by flyinghead** [comment]
At least we know the issue started on or before 81842cc (2024-07-03T15:43:49.000Z)
Thanks

**Comment by duke912** [comment]
Can confirm the graphical bug still persist on the latest retroarch version when running on vulkan api.

**Comment by efrainm23** [comment]
The issue also happens in Retroarch for me regardless of full framebuffer emulation being on or not. I have always run with native rendering resolution. DX11 or the Retroarch glcore renderer work fine in Retroarch on an Nvidia card. 

As far as I can tell from their commits the Flycast core they have listed there has not been updated for a long time at [https://github.com/libretro/flycast](url). I may be looking at the wrong repository. Maybe this one is the one they derive their core from and syncs to upstream. 

This does seem to indicate something caused by a driver that has affected the Flycast Vulkan renderer from way back. It seems that some people have had luck with older versions of Retroarch before 1.19. Not sure how that could have an effect if it apparently  had existed in standalone as of the last post in this thread in November. Maybe the older versions of Retroarch are using an older Vulkan driver that does not trigger the issue.

**Comment by ToniBC** [comment]
It's not a Retroarch problem, the same thing has been happening in Flycast standalone nightlies for a while now. Vulkan generates these graphical errors. At least with RTX30 graphics and Windows.

This is rarely reported, because by default, Retroarch uses DX11, which is the same driver that Flycast uses by default, so users, as they don't change the video driver, don't notice it.

There are no problems in DX11.

PS: The repository is maintained from the official one, not from Libretro's, that's why you don't see any movements.

**Comment by dantesquivel** [comment]
> Can you try Flycast standalone?

I'm experiencing this same issue on standalone. Win 11, i9-12900K, RTX 3080 Ti on latest drivers. Yup don't even need a game. Boot the DC bios, go to the settings choose language so the language options appear and you'll see the artifactimg whenever Vulkan per pixel is selected. The problem can be improved by lowering the resolution scale but it's not fully eliminated. Running latest Dev version of flycast.

Somewhat related, this specific menu is actually a go to for me to test Flycast. On my RetroidPocket Mini this menu shows me I cannot run DC on anything but native resolution with per pixel on Vulkan, it doesn't artifact in this scenario but the frame rate clearly dips.

**Comment by dantesquivel** [comment]
> The corruption appears even in the Flycast 2.3.2-199 nightly and in Retroarch with the Vulkan Driver. Even in the BIOS as they say.
> 
> With DX11 it seems to be mitigated or it doesn't appear, the same if you set the native resolution, it doesn't appear in the BIOS, but if you increase the resolution, there are artifacts in the BIOS at the bottom right, above the clock icon, both in Flycast and Retroarch.
> 
> If Full Framebuffer Emulation is activated, the artifacts no longer appear, but we can't increase the resolution.
> 
> This has been happening for a while and as it happens in both, it's not focused on AR.
> 
> Many won't notice it because they use old versions, like March 2.3, but it seems to be something related to the framebuffer.



> It's not a Retroarch problem, the same thing has been happening in Flycast standalone nightlies for a while now. Vulkan generates these graphical errors. At least with RTX30 graphics and Windows.
> 
> This is rarely reported, because by default, Retroarch uses DX11, which is the same driver that Flycast uses by default, so users, as they don't change the video driver, don't notice it.
> 
> There are no problems in DX11.
> 
> PS: The repository is maintained from the official one, not from Libretro's, that's why you don't see any movements.

Hi there, experiencing the exact same issue. Also with a fairly beefy PC with ample resources. Will update here if I find anything else.

**Comment by Adad-G** [comment]
Same here on Vulkan, even on the standalone version: RetroArch 1.21.0, Windows 11, 5600X, RTX 3070 (576.80).

**Comment by bslenul** [comment]
I'm experiencing this now, I haven't used Flycast in a while + I updated quite a few things lately (BIOS, Nvidia drivers, Windwows 10 -> 11 with all the Windows updates) so I have absolutely no idea what triggered it :/ 

Issue happens in both RetroArch and standalone, much more noticeable with RetroArch tho.

<img width="1282" height="752" alt="Image" src="https://github.com/user-attachments/assets/f6986643-82eb-46aa-be3c-8d2f4017c9e4" />

<img width="918" height="739" alt="Image" src="https://github.com/user-attachments/assets/c2f0d4fc-028e-4fdc-88d6-1a05da1d1aef" />

**Comment by flyinghead** [comment]
I suspect an nvidia driver issue. And it also happens with the per triangle/per strip modes, which are very basic and don't use any advanced Vulkan feature.

**Comment by flyinghead** [comment]
Upgraded to driver 575.64 and I still can't reproduce it.

**Comment by bslenul** [comment]
FWIW I'm using 580.97 and I'm using a 3060 Ti.
Also idk how it works internally so it might be expected, but the issue doesn't seem to happen when "Full Framebuffer Emulation" is enabled.

**Comment by flyinghead** [comment]
All reports mention RTX 30xx gpus and I have a 1080.

**Comment by bslenul** [comment]
Had some time to check this a bit more, I was able to bisect a commit from RetroArch: https://github.com/libretro/RetroArch/commit/fed2e10d9712d6e2082c99420fe0b7b483379c1a

No idea if that makes sense to you, but the glitch happen with that commit and no black squares on the previous commit:

https://github.com/user-attachments/assets/4119bc0d-926e-476a-a764-df5137567f4c

**Comment by bslenul** [comment]
Will try to bisect standalone later (no promises tho, last time I tried to bisect I was having compilation errors with older commits :/) but I grabbed older versions from the Releases page, can't spot any glitches with 2.3 (966ff3767) but they're present in 2.4 (8108e6390).

**Comment by bslenul** [comment]
Seems to have started with cc56b0fbaef0a818a2f9a361fc1f59e3c5d68abf, previous commit (5755c5b3c6c2d08e3660e38380be996b8727dd66) vs. the "bad" one:

https://github.com/user-attachments/assets/146c3b40-4b34-4b15-ac43-1c4301377558

**edit:** Hm, to be completely fair, from time to time the black squares seem to pop even with the "good" commit, it only pops for a frame so it's nowhere near as bad as the next commits and it doesn't even happen consistently, sometimes I'm able to boot the game 10x in a row without it happening...

https://github.com/user-attachments/assets/1cdc68a2-6499-4bf1-a112-f1dac2c62c48

Still I thought it was worth mentioning, maybe the "bad" commit isn't really the cause but only made the issue more visible.

**Comment by flyinghead** [comment]
Great job and thank you for your help!
This commit (https://github.com/flyinghead/flycast/commit/cc56b0fbaef0a818a2f9a361fc1f59e3c5d68abf) was done to support multiple render calls for a single frame. As a result, the vulkan render pass is kept open as long as possible, and command buffers aren't submitted immediately, so that a second render can take place before the frame is presented.

It looks like the RTX 30xx GPU (or driver) is having synchronization issues, and this change makes it worse. I believe the code is correct and I checked with the validation layers that no error or warning was being raised when I worked on this issue. But of course I can't rule out a bug in Flycast.

Did you try to run the vulkan SDK validation layers? This could give us a clue about the issue if any, and/or help find a workaround.

**Comment by bslenul** [comment]
I had an older Vulkan SDK installed and for some reason no glitches while it was attached to Flycast process 🤔 
I installed the latest one and the glitches appear but no validation layers errors, if I select Synchronization however in the "Apply a Vulkan Loader Configuration" however I'm getting errors: [flycast.txt](https://github.com/user-attachments/files/22208254/flycast.txt)

I honestly don't really know how to use this stuff, no idea if this helps or not :p

**Comment by efrainm23** [comment]
I tried the Retroarch core again with Vulkan for the first time in a while and I found it is actually working properly now on my RTX 5080. I had experienced this issue back in February on Vulkan shortly after having gotten the new GPU. I have not tried standalone. I do not have full framebuffer emulation on, but I do have per-pixel sorting on.

**Comment by dantesquivel** [comment]
> I had an older Vulkan SDK installed and for some reason no glitches while it was attached to Flycast process 🤔 I installed the latest one and the glitches appear but no validation layers errors, if I select Synchronization however in the "Apply a Vulkan Loader Configuration" however I'm getting errors: [flycast.txt](https://github.com/user-attachments/files/22208254/flycast.txt)
> 
> I honestly don't really know how to use this stuff, no idea if this helps or not :p

Did you ever figure this out? 
I'm on the latest nvidia drivers on a 3080Ti and getting this same artifacting specifically with Vulkan on per pixel
I'm  relegated to DirectX11 per pixel to avoid the problem

EDIT: Confirming that this is related to an Nvidia issue in RTX 30xx cards:
https://www.reddit.com/r/EVGA/comments/wkf9mf/rtx_3000_series_artifacts_black_squares_that/

**Comment by flyinghead** [comment]
Pushed a tentative fix on the `dev` branch, for both standalone and the libretro core.

**Comment by bslenul** [comment]
The fix is working great for me, on both standalone and the Libretro core 👍 If anyone wants to try the core (not yet available on the Online Updater): https://github.com/user-attachments/files/22871182/flycast_libretro.zip

**Comment by PoloniumRain** [comment]
@flyinghead @bslenul Fix is working well in both standalone and RetroArch! Couldn't find any graphical corruption with 10 games tested in both of them.

In my original post i only mentioned Vulkan, but since then i noticed the same issue when using D3D11 in RA as well, and that's also fixed now. So both Vulkan and D3D11 are working fine, with or without shaders. It's been just over a year now and it looks like this issue is finally fixed :)

I'll leave this issue open incase anyone else wants to leave feedback.

**Comment by flyinghead** [comment]
merged into the master branch so should be available through the retroarch online updater soon.

**Comment by dantesquivel** [comment]
> merged into the master branch so should be available through the retroarch online updater soon.

This worked!!
